mixer/adapter/svcctrl/svcctrl_test.go

```diff
@@ -17,49 +17,162 @@ package svcctrl
 import (
 	"reflect"
 	"testing"
-	"time"
 
-	"github.com/davecgh/go-spew/spew"
-	sc "google.golang.org/api/servicecontrol/v1"
-)
+	pbtypes "github.com/gogo/protobuf/types"
 
-func TestHandleMetric(t *testing.T) {
-	timeNow := time.Now().Format(time.RFC3339Nano)
-	request, err := handleMetric(timeNow, "")
+	"istio.io/istio/mixer/adapter/svcctrl/config"
+	"istio.io/istio/mixer/adapter/svcctrl/template/svcctrlreport"
+	"istio.io/istio/mixer/pkg/adapter"
+	at "istio.io/istio/mixer/pkg/adapter/test"
+	"istio.io/istio/mixer/template/apikey"
+	"istio.io/istio/mixer/template/quota"
+)
 
+func TestInitializeHandlerContext(t *testing.T) {
+	adapterCfg := getTestAdapterConfig()
+	ctx, err := initializeHandlerContext(at.NewEnv(t), adapterCfg, &mockSvcctrlClient{})
 	if err != nil {
-		t.Fatalf("handleMetric() failed with:%v", err)
+		t.Errorf("initializeHandlerContext() failed with %v", err)
+	}
+
+	expectedIdx := map[string]*config.ServiceConfig{
+		"service_a": adapterCfg.ServiceConfigs[0],
+		"service_b": adapterCfg.ServiceConfigs[1],
+	}
+	if !reflect.DeepEqual(expectedIdx, ctx.serviceConfigIndex) {
+		t.Errorf("expect serviceConfigIndex :%v, but get %v",
+			expectedIdx, ctx.serviceConfigIndex)
+	}
+}
+
+func TestConfigValidation(t *testing.T) {
+
+	{
+		b := getTestBuilder()
+		err := b.Validate()
+		if err != nil {
+			t.Errorf(`valid config, but get error %v`, err.Multi)
+		}
+	}
+
+	invalidBuilders := []*builder{
+		func() *builder {
+			b := getTestBuilder()
+			b.config.RuntimeConfig = nil
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			b.config.ServiceConfigs = []*config.ServiceConfig{}
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			b.config.ServiceConfigs = nil
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			b.config.ServiceConfigs[0].MeshServiceName = ""
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			b.config.ServiceConfigs[0].GoogleServiceName = ""
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			b.config.ServiceConfigs[0].QuotaConfigs[0].QuotaName = ""
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			b.config.ServiceConfigs[0].QuotaConfigs[0].Expiration = nil
+			return b
+		}(),
+		func() *builder {
+			b := getTestBuilder()
+			expiration := b.config.ServiceConfigs[0].QuotaConfigs[0].Expiration
+			expiration.Seconds = 0
+			expiration.Nanos = 0
+			return b
+		}(),
+	}
+
+	for _, b := range invalidBuilders {
+		err := b.Validate()
+		if err == nil {
+			t.Errorf(`fail to detect invalid config: %v`, *b.config)
+		}
+	}
+}
+
+func TestBuilderInterface(t *testing.T) {
+	var b interface{} = &builder{}
+	_, ok := b.(adapter.HandlerBuilder)
```

Comment:
instead of making this a test, make these part of the code. catch it at compile time.